### PR TITLE
ASN.1 Parsing

### PR DESCRIPTION
**Pull request by bdehamer**
#### Summary
Adds support for basic parsing of DER-encoded ASN.1 objects. This lays the groundwork for adding full x509 certificate support so that we can do verification of the Fulcio certificate chain.

This implementation does not support the complete set of ANS.1 datatypes or encoding options. For now, it is focused exclusively on the ASN.1 types commonly used in x509 certificates using DER encoding:

* `BOOLEAN`
* `INTEGER`
* `BIT STRING`
* `OCTET STRING`
* `OBJECT IDENTIFER`
* `SEQUENCE`
* `SET`  
* `PrintableString`
* `UTCTime`

Following is the structure resulting from parsing a Fulcio-issued x509 signing certificate:
```
SEQUENCE (3 elem) 
  SEQUENCE (8 elem) 
    [0] (1 elem) 
      INTEGER (1 byte) 2
    INTEGER (20 byte) 651968271402900508168874304870743208745138152437
    SEQUENCE (1 elem) 
      OBJECT IDENTIFIER 1.2.840.10045.4.3.3
    SEQUENCE (2 elem) 
      SET (1 elem) 
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 2.5.4.10
          PrintableString sigstore.dev
      SET (1 elem) 
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 2.5.4.3
          PrintableString sigstore-intermediate
    SEQUENCE (2 elem) 
      UTCTime Wed, 21 Sep 2022 19:25:15 GMT
      UTCTime Wed, 21 Sep 2022 19:35:15 GMT
    SEQUENCE (0 elem) 
    SEQUENCE (2 elem) 
      SEQUENCE (2 elem) 
        OBJECT IDENTIFIER 1.2.840.10045.2.1
        OBJECT IDENTIFIER 1.2.840.10045.3.1.7
      BIT STRING (520 bit) 000001001111000010101100000011010000101001001000101000100011001010001...
    [3] (1 elem) 
      SEQUENCE (12 elem) 
        SEQUENCE (3 elem) 
          OBJECT IDENTIFIER 2.5.29.15
          BOOLEAN true
          OCTET STRING (4 byte) 03020780
            BIT STRING (1 bit) 1
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 2.5.29.37
          OCTET STRING (12 byte) 300A06082B06010505070303
            SEQUENCE (1 elem) 
              OBJECT IDENTIFIER 1.3.6.1.5.5.7.3.3
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 2.5.29.14
          OCTET STRING (22 byte) 041476639CDB701727D02F5E5EBD5AAB63D98CFEAE5E
            OCTET STRING (20 byte) 76639CDB701727D02F5E5EBD5AAB63D98CFEAE5E
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 2.5.29.35
          OCTET STRING (24 byte) 30168014DFD3E9CF56241196F9A8D8E92855A2C62E18643F
            SEQUENCE (1 elem) 
              [0] (20 byte) DFD3E9CF56241196F9A8D8E92855A2C62E18643F
        SEQUENCE (3 elem) 
          OBJECT IDENTIFIER 2.5.29.17
          BOOLEAN true
          OCTET STRING (100 byte) 3062866068747470733A2F2F6769746875622E636F6D2F6769746875622F747275737...
            SEQUENCE (1 elem) 
              [6] (96 byte) https://github.com/github/trust-metadata-demo/.github/workflows/npm-demo.yml@refs/heads/npm-demo
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.57264.1.1
          OCTET STRING (43 byte) https://token.actions.githubusercontent.com
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.57264.1.2
          OCTET STRING (4 byte) push
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.57264.1.3
          OCTET STRING (40 byte) 221b26650574313a4207ae9262c6ce0707859a39
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.57264.1.4
          OCTET STRING (8 byte) npm-demo
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.57264.1.5
          OCTET STRING (26 byte) github/trust-metadata-demo
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.57264.1.6
          OCTET STRING (19 byte) refs/heads/npm-demo
        SEQUENCE (2 elem) 
          OBJECT IDENTIFIER 1.3.6.1.4.1.11129.2.4.2
          OCTET STRING (123 byte) 04790077007500086092F02852FF6845D1D16B27849C456718AC163DC338D26DE6BC2...
            OCTET STRING (121 byte) 0077007500086092F02852FF6845D1D16B27849C456718AC163DC338D26DE6BC22063...
  SEQUENCE (1 elem) 
    OBJECT IDENTIFIER 1.2.840.10045.4.3.3
  BIT STRING (824 bit) 001100000110010100000010001100000000010111000111111011100010010111000...
```

### CODE REVIEW

File: src/x509/asn1/parse.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    : RE_TIME_LONG_YEAR.exec(timeStr);
+
+  if (!m) {
+    throw new Error('invalid time');
+  }
+
+  // Normalize to 4-digit year
+  if (shortYear) {
+    let year = Number(m[1]);
+    year += year >= 50 ? 1900 : 2000;
```
Comment (feelepxyz) [inline src/x509/asn1/parse.ts:58]: Presuming this is for legacy/ancient certificates? Maybe add a comment for future selves?

File: src/x509/asn1/parse.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  // Parse the time string into matches - captured groups start at index 1
+  const m = shortYear
+    ? RE_TIME_SHORT_YEAR.exec(timeStr)
+    : RE_TIME_LONG_YEAR.exec(timeStr);
+
+  if (!m) {
+    throw new Error('invalid time');
+  }
+
+  // Normalize to 4-digit year
```
Comment (feelepxyz) [inline src/x509/asn1/parse.ts:55]:
```suggestion
  // Normalize to 4-digit year (handle legacy certificates with only YY)
```

File: src/__tests__/x509/asn1/obj.test.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        expect(obj).toBeInstanceOf(ASN1Obj);
+        expect(obj.tag.number).toBe(2);
+        expect(obj.value.toString('hex')).toBe('1010');
+        expect(obj.subs).toHaveLength(0);
+        expect(obj.raw).toEqual(buffer);
+      });
+    });
+
+    describe('when parsing a constructed object', () => {
+      const buffer = Buffer.from('30080202101002021111', 'hex');
```
Comment (feelepxyz) [inline src/__tests__/x509/asn1/obj.test.ts:21]: ooc where are these values coming from?

**Review by feelepxyz — DISMISSED** [review]
Test coverage looks good and the reference comments are helpful 👍  I can't say that I've in detail reviewed all the bit operations but looks good to me. Also tagged Fredrik in case he might have some experience parsing asn.1/der things.

File: src/__tests__/x509/asn1/length.test.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      0x87, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    ]);
+    it('throws an error', () => {
+      expect(() => decodeLength(stream)).toThrowError(
+        'length exceeds 6 byte limit'
+      );
+    });
+  });
+
+  describe('when the encoded length is indefinite', () => {
```
Comment (kommendorkapten) [inline src/__tests__/x509/asn1/length.test.ts:37]: Not sure if this is a relevant comment, but IIRC "infinite encoding" can be used even if the length is small, the length just happens to be unknown during encoding. But if we have not yet seen any examples of this, we can punt an implementation of it.

File: src/__tests__/x509/asn1/parse.test.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    expect(parseBoolean(Buffer.from([0xff]))).toEqual(true);
+  });
+});
+
+describe('parseBitString', () => {
+  it('parses bit strings', () => {
+    expect(parseBitString(Buffer.from([0x04, 0x57, 0xd0]))).toEqual([
+      0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1,
+    ]);
+    expect(parseBitString(Buffer.from([0x04, 0x57, 0xdf]))).toEqual([
```
Comment (kommendorkapten) [inline src/__tests__/x509/asn1/parse.test.ts:115]: Maybe add a comment here that the four last bits are unused and so doesn't affect the parsed value?

File: src/x509/asn1/length.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+import { ByteStream } from '../../util/stream';
+import { ASN1ParseError } from './error';
+
+// Decodes the length of a DER-encoded ANS.1 element from the supplied stream.
+// https://learn.microsoft.com/en-us/windows/win32/seccertenroll/about-encoded-length-and-value-bytes
+export function decodeLength(stream: ByteStream): number {
+  const buf = stream.get();
+
+  // If the first bit is unset the length is just the value of the byte.
```
Comment (kommendorkapten) [inline src/x509/asn1/length.ts:11]:
```suggestion
  // If the most significant bit is unset the length is just the value of the byte.
```

**Comment by kommendorkapten** [comment]
Very good job 🚀 
Some minor comments only.

File: src/x509/asn1/parse.ts (hunk 1 above)
Comment (bdehamer) [reply src/x509/asn1/parse.ts:58]: Yeah, this is the scheme described in the spec (https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.5.1). Will add a comment.

File: src/__tests__/x509/asn1/obj.test.ts (hunk 3 above)
Comment (bdehamer) [reply src/__tests__/x509/asn1/obj.test.ts:21]: Just some test data I invented to exercise all the various parsing paths. I added some comments which better explain what the data represents.

File: src/__tests__/x509/asn1/length.test.ts (hunk 4 above)
Comment (bdehamer) [reply src/__tests__/x509/asn1/length.test.ts:37]: I didn't see the indefinite length encoding used in any of the certs we're handling so I decided not to support it (makes parsing a bit more complicated).

File: src/__tests__/x509/asn1/parse.test.ts (hunk 5 above)
Comment (bdehamer) [reply src/__tests__/x509/asn1/parse.test.ts:115]: 👍

File: src/x509/asn1/length.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  if (buf < 0x80) {
```
[the remark this answers is not in the document]
Comment (bdehamer) [reply src/x509/asn1/length.ts:12]: Yeah, I struggled to find the most "readable" option for some of these bit-wise operations. Using a bitwise-and is probably a more idiomatic way to do a is-this-bit-set check so I'll change to that.

I think the correct expression would actually be:
```js
if ((buf & 0x80) === 0x00)
```

**Comment by bdehamer** [comment]
Added some more detailed comments based on feedback from @feelepxyz / @kommendorkapten

**Review by kommendorkapten — APPROVED** [review]
🚀

**Review by feelepxyz — APPROVED** [review]
💯 :shipit:

**Pull request merged**